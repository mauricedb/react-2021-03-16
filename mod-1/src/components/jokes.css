.jokes {
  display: grid;
  grid-gap: 1rem;
  grid-template-columns: 1fr 1fr;
  margin: 1rem;
}

.panel {
  border: 1px solid lightgray;
  padding: 1rem;
}
