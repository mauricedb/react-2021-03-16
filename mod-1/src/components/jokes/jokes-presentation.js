import React, { Fragment, useState, useMemo } from 'react';
import propTypes from 'prop-types';

import { SelectedJoke } from './selected-joke';

export default function D({ heading, jokes, setSelected, selected }) {
  const [ratingToShow] = useState(4);
  const goodJokes = useMemo(
    () => jokes.filter((j) => j.rating > ratingToShow),
    [jokes, ratingToShow]
  );

  return (
    <div>
      <h1>{heading}</h1>

      <div className="jokes">
        <div className="panel">
          <ul>
            {jokes.map((j) => (
              <Fragment key={j.id}>
                <li onClick={() => setSelected(j)}>{j.joke}</li>
                <li> Stuff</li>
              </Fragment>
            ))}
          </ul>
        </div>
        <div className="panel">
          {selected && <SelectedJoke key={selected.id} selected={selected} />}
        </div>
      </div>
    </div>
  );
}

D.displayName = 'The component name';

D.propTypes = {
  selected: propTypes.object,
};

D.defaultProps = {
  heading: 'Some jokes',
  selected: {
    id: 1,
    joke: 'Not yet',
  },
};
