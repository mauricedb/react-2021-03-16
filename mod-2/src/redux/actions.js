export const actionTypes = {
  loadJokes: 'load-jokes',
  deleteJoke: 'delete-joke',
};

export function deleteJoke(id) {
  return { type: actionTypes.deleteJoke, payload: id };
}

export function loadJokes(jokes) {
  if (!Array.isArray(jokes)) {
    throw new Error('Jokes should be an array');
  }
  return { type: actionTypes.loadJokes, payload: jokes };
}

export function fetchJokes() {
  return async function worker(dispatch) {
    const rsp = await fetch(
      'https://the-problem-solver-sample-data.azurewebsites.net/jon-skeet'
    );
    const jokes = await rsp.json();
    dispatch(loadJokes(jokes));
  };
}
