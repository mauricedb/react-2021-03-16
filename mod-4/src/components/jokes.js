import React, { useEffect, useState } from 'react';
import classNames from 'classnames';

import classes from './jokes.module.css';

import { JokeDetails } from './joke-details';
import { SingleJoke } from './single-joke';
import { ErrorBoundary } from './error-boundary';

import { useFetch } from './useFetch';
import { themeContext } from './theme-context';

function ErrorDisplay({ error }) {
  return <div>There was an error: {error.message}</div>;
}

export function Jokes({ heading, url }) {
  const { theme } = React.useContext(themeContext);
  const [selected, setSelected] = useState(null);

  const { data: jokes, status } = useFetch(url);

  if (status === 'error') {
    return <div>Some error...</div>;
  }

  if (status !== 'ready') {
    return <div>Loading...</div>;
  }

  return (
    <div>
      <h1>{heading}</h1>

      <div className={classes.jokes}>
        <div className={classNames(classes.panel, classes[theme])}>
          <ul>
            {jokes.map((joke) => (
              <SingleJoke key={joke.id} joke={joke} setSelected={setSelected} />
            ))}
          </ul>
        </div>
        <div className={classNames(classes.panel, classes[theme])}>
          <ErrorBoundary key={selected?.id} errorComponent={<ErrorDisplay />}>
            {() => <JokeDetails joke={selected} />}
          </ErrorBoundary>
        </div>
      </div>
    </div>
  );
}
