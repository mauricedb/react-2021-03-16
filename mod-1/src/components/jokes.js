import React, { useEffect, useState } from 'react';
import './jokes.css';

import { rating } from './utils';

const url =
  'https://the-problem-solver-sample-data.azurewebsites.net/jon-skeet';

export function Jokes({ heading }) {
  const [selected, setSelected] = useState(null);
  const [jokes, setJokes] = useState(null);

  useEffect(() => {
    async function fetchJokes() {
      const rsp = await fetch(url);
      const jokes = await rsp.json();
      setJokes(jokes);
    }

    fetchJokes();
  }, []);

  if (!jokes) {
    return <div>Loading...</div>;
  }

  return (
    <div>
      <h1>{heading}</h1>

      <div className="jokes">
        <div className="panel">
          <ul>
            {jokes.map((j) => (
              <li onClick={() => setSelected(j)}>{j.joke}</li>
            ))}
          </ul>
        </div>
        <div className="panel">
          {selected && (
            <div>
              <p>
                <b>Joke:</b> {selected.joke}
              </p>
              <p>
                <b>Rating:</b> {rating(selected)}
              </p>
            </div>
          )}
        </div>
      </div>
    </div>
  );
}
