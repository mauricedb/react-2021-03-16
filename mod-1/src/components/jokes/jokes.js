import React, { useEffect, useState, Fragment } from 'react';
import './jokes.css';

import D from './jokes-presentation';

const url =
  'https://the-problem-solver-sample-data.azurewebsites.net/jon-skeet';

export function Jokes({ heading }) {
  // useState(heading)
  const [selected, setSelected] = useState(null);
  const [jokes, setJokes] = useState(null);

  useEffect(() => {
    async function fetchJokes() {
      const rsp = await fetch(url);
      const jokes = await rsp.json();
      setJokes(jokes);
    }

    fetchJokes();
  }, []);

  if (!jokes) {
    return <div>Loading...</div>;
  }

  return (
    <D
      heading={heading}
      jokes={jokes}
      selected={selected}
      setSelected={setSelected}
    />
  );
}
