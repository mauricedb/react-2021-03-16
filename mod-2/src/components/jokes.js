import React from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { deleteJoke } from '../redux/actions';

export function Jokes() {
  const dispatch = useDispatch();
  const jokes = useSelector((state) => state.jokes);

  return (
    <div>
      <h2>Chuck Norris Jokes</h2>
      <ul>
        {jokes.map((joke) => (
          <li key={joke.id}>
            {joke.joke}{' '}
            <button onClick={() => dispatch(deleteJoke(joke.id))}>
              &#9249;
            </button>
          </li>
        ))}
      </ul>
    </div>
  );
}
