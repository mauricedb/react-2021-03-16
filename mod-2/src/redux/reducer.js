import { actionTypes } from './actions';

export function reducer(state = { jokes: [] }, action) {
  switch (action.type) {
    case actionTypes.loadJokes:
      return { ...state, jokes: action.payload };
    case actionTypes.deleteJoke:
      return {
        ...state,
        jokes: state.jokes.filter((joke) => joke.id !== action.payload),
      };
    default:
      return state;
  }
}
