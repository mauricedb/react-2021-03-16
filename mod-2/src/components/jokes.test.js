import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { render, screen } from '@testing-library/react';
import { Jokes } from './jokes';

function renderJokes(jokes) {
  const store = createStore(() => ({ jokes }));
  return render(
    <Provider store={store}>
      <Jokes />
    </Provider>
  );
}

test('renders no jokes', () => {
  renderJokes([]);
  screen.debug();
});

test('renders one joke', () => {
  renderJokes([{ id: 1, joke: 'Some joke' }]);
  screen.debug();
});

test('renders two joke', () => {
  const { container } = renderJokes([
    { id: 1, joke: 'Some joke' },
    { id: 2, joke: 'Another joke' },
  ]);

  expect(container).toMatchSnapshot();
});
