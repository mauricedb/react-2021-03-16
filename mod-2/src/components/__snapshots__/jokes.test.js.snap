// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders two joke 1`] = `
<div>
  <div>
    <h2>
      Chuck Norris great Jokes
    </h2>
    <ul>
      <li>
        Some joke
         
        <button>
          ␡
        </button>
      </li>
      <li>
        Another joke
         
        <button>
          ␡
        </button>
      </li>
    </ul>
  </div>
</div>
`;
